src/network.rs

```diff
@@ -87,140 +95,80 @@ impl Network {
-        info!(
-            "Random Adds: {} Drops: {}",
-            log::important(adds),
-            log::important(drops)
-        );
-    }
-
-    fn handle_messages(&mut self) -> TickStats {
-        let mut responses = Vec::new();
+    fn handle_actions(&mut self, actions: &mut Vec<Action>) -> TickStats {
         let mut stats = TickStats::new();
 
-        loop {
-            for section in self.sections.values_mut() {
-                responses.extend(section.handle_requests(&self.params));
-            }
-
-            if responses.is_empty() {
-                break;
-            }
-
-            stats += self.handle_responses(&mut responses)
-        }
+        for action in actions.drain(..) {
+            match action {
+                Action::Reject(_) => {
+                    stats.rejections += 1;
+                }
+                Action::Merge(target) => {
+                    let sources: Vec<_> = self.sections
+                        .keys()
+                        .filter(|prefix| prefix.is_descendant(&target))
+                        .cloned()
+                        .collect();
+
+                    if sources.is_empty() {
+                        continue;
+                    }
 
-        stats
-    }
+                    let sources: Vec<_> = sources
+                        .into_iter()
+                        .map(|source| self.sections.remove(&source).unwrap())
+                        .collect();
 
-    fn handle_responses(&mut self, responses: &mut Vec<Response>) -> TickStats {
-        let mut stats = TickStats::new();
+                    stats.merges += 1;
 
-        for response in responses.drain(..) {
-            match response {
-                Response::Merge(section, old_prefix) => {
-                    self.sections
-                        .entry(section.prefix())
-                        .or_insert_with(|| {
-                            stats.merges += 1;
-                            Section::new(section.prefix())
-                        })
-                        .merge(&self.params, section);
-                    let _ = self.sections.remove(&old_prefix);
+                    let section = self.sections.entry(target).or_insert_with(
+                        || Section::new(target),
+                    );
+                    for source in sources {
+                        section.merge(&self.params, source);
+                    }
                 }
-                Response::Split(section0, section1, old_prefix) => {
+                Action::Split(source) => {
                     stats.splits += 1;
 
-                    let prefix0 = section0.prefix();
-                    let prefix1 = section1.prefix();
+                    let source = if let Some(section) = self.sections.remove(&source) {
+                        section
+                    } else {
+                        // Pre-split section does not exists due to being already merged
+                        // or split.
```

Comment:
there shall be only one action::split for a section, so how could the section be split before handling that action?

Reply:
Think about scenario like this:
1. A node joins a section. Section decides it want's to split, so it emits `Action::Split`.
2. Before that action is handled, the same section is joined by another node, which also triggers split.
3. The first `Split` is handled, removing the original section from the list of section and inserting the post-split sections.
4. The second `Split` is now handled too, but the pre-split section is no longer there.

In the past I was explicitly handling this situation by having the `state` flag in `Section` and checking whether section is already splitting or merging. I realised that it's not really necessary, because the section, even if it decided to split or merge, can still keep handing all the messages as usual, and when it actually splits (or merges) all the data in the section (nodes, chain, caches, messages, ...) will be sorted to the post-split / post-merge section accordingly and then handled normally. This simplifies things. The only issue is that now the `Split` / `Merge` actions can arrive multiple times, but I believe it's OK to just ignore them.

Reply:
regarding step 2 `Before that action is handled, the same section is joined by another node, which also triggers split.`